### PR TITLE
🏗️ Compound Interest Accumulation and Rate Rebalancing 🚀 

### PR DESCRIPTION
**Pull request by abigger87**

### CODE REVIEW

File: src/CharityVaultFactory.sol [hunk 1]
```diff
@@ -80,11 +84,10 @@ contract CharityVaultFactory {
                 address(this),
                 // Compute Inline CharityVault Salt, h/t @t11s
                 keccak256(
-                    abi.encodePacked(
-                        address(underlying),
+                    abi.encode(
```
Comment (transmissions11) [inline src/CharityVaultFactory.sol:87]: you use abi.encodePacked for computing above, you need to match here

File: src/CharityVaultFactory.sol [hunk 2]
```diff
@@ -80,11 +84,10 @@ contract CharityVaultFactory {
...
+                        address(underlying).fillLast12Bytes(),
```
Comment (transmissions11) [inline src/CharityVaultFactory.sol:88]: you don't do this above so need to match here

File: src/CharityVaultFactory.sol [hunk 3]
```diff
@@ -52,10 +56,10 @@ contract CharityVaultFactory {
                     address(underlying),
                     charity,
                     feePercent,
-                    address(vaultFactory.getVaultFromUnderlying(underlying))
+                    address(VAULT_FACTORY.getVaultFromUnderlying(underlying))
```
Comment (transmissions11) [inline src/CharityVaultFactory.sol:59]: ya dont need this

File: src/CharityVault.sol [hunk 4]
```diff
@@ -89,51 +103,79 @@ contract CharityVault is ERC20, Auth {
...
 
-        // Transfer in underlying tokens from the sender to the vault
-        underlying.safeTransferFrom(msg.sender, address(vault), underlyingAmount);
+        // Deposit to the VAULT
+        VAULT.deposit(underlyingAmount);
+
+        // Determine the equivalent amount of rcvTokens and mint them.
+        _mint(msg.sender, underlyingAmount.fdiv(exchangeRate(), BASE_UNIT));
 
         emit CharityDeposit(msg.sender, underlyingAmount);
```
Comment (transmissions11) [inline src/CharityVault.sol:139]: you open urself up to reentrancy by doing the mint after external interactions

File: src/CharityVault.sol [hunk 5]
```diff
@@ -89,51 +103,79 @@ contract CharityVault is ERC20, Auth {
...
+
+        // Transfer underlying tokens to the user.
+        UNDERLYING.safeTransfer(msg.sender, underlyingAmount);
+
+        // Determine the equivalent amount of rcvTokens and burn them.
+        // This will revert if the user does not have enough rcvTokens.
+        _burn(msg.sender, underlyingAmount.fdiv(exchangeRate(), BASE_UNIT));
+
+        emit CharityWithdraw(msg.sender, underlyingAmount);
+
```
Comment (transmissions11) [inline src/CharityVault.sol:160]: ibid

File: src/CharityVault.sol [hunk 6]
```diff
@@ -143,96 +185,166 @@ contract CharityVault is ERC20, Auth {
...
+            assembly {
+                calldatacopy(0, 0, calldatasize())
+                let result := delegatecall(gas(), payable(address(VAULT)), 0, calldatasize(), 0, 0)
+                returndatacopy(0, 0, returndatasize())
+                switch result
+                case 0 { revert(0, returndatasize()) }
+                default { return(0, returndatasize()) }
+            }
+        }
 }
```
Comment (transmissions11) [inline src/CharityVault.sol:350]: lol interesting but why would people want this? feels unnecessarily risky

File: src/CharityVault.sol [hunk 7]
```diff
@@ -143,96 +185,166 @@ contract CharityVault is ERC20, Auth {
...
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Erroneous ether sent will be forward to the charity as a donation
+    receive() external payable {
+        (bool sent, bytes memory _data) = CHARITY.call{value: msg.value}("");
+        require(sent, "Failed to send to CHARITY");
 
+        // If sent, emit logging event
+        emit TransparentTransfer(msg.sender, msg.value);
+    }
```
Comment (transmissions11) [inline src/CharityVault.sol:335]: this will break withdrawing an ETH vault

File: src/CharityVault.sol [hunk 8]
```diff
@@ -143,96 +185,166 @@ contract CharityVault is ERC20, Auth {
...
 
-        // Burn inputed fcvTokens.
-        _burn(charity, amount);
+        // Burn inputed rcvTokens.
+        _burn(CHARITY, amount);
 
-        // If the withdrawal amount is greater than the float, pull tokens from Fuse.
-        // if (underlyingAmount > getFloat()) vault.pullIntoFloat(underlyingAmount);
+        // Get the user's underlying balance
+        uint256 user_balance = VAULT.balanceOfUnderlying(msg.sender);
```
Comment (transmissions11) [inline src/CharityVault.sol:211]: idt u want the user's balance here the vault is custodying vault shares on behalf of the them

File: src/CharityVault.sol (hunk 6 above)
Comment (abigger87) [reply src/CharityVault.sol:350]: haha just me experimenting

Comment (abigger87) [reply src/CharityVault.sol:350]: will remove

File: src/CharityVault.sol (hunk 5 above)
Comment (abigger87) [reply src/CharityVault.sol:160]: idgi @transmissions11

Comment (transmissions11) [reply src/CharityVault.sol:160]: ibid == same as above

**Comment by abigger87** [comment]
merging for sync

**Pull request merged**